Content.Server/Cargo/Systems/CargoTrackingSystem.cs

```diff
@@ -0,0 +1,195 @@
+        if (!TryComp<StationCargoOrderDatabaseComponent>(station, out var cargoOrderDatabaseComponent))
+            return;
+
+        // first check if the order is still in the order list
+        foreach (var order in cargoOrderDatabaseComponent.Orders)
+        {
+            if (order.OrderId == cargoInvoiceComponent.OrderId)
+            {
+                // the order is still out for delivery, so we start waiting
+                _popupSystem.PopupEntity(Loc.GetString("cargo-tracker-not-arrived"), uid, args.User);
+                StartWaiting(uid, component);
+                return;
+            }
+        }
+
+        // lookup the order from the list
+        if (cargoOrderDatabaseComponent.OrderLookup.TryGetValue(component.TrackedOrderId.Value, out var orderUid))
+        {
+            // start tracking the order
+            StartTrackingArrival(uid, orderUid, component, args.User);
+            return;
+        }
+
+        // if you've gotten this far, yet the order is no longer in the order list or order lookup, it must mean that the tracked entity has been destroyed
+        _popupSystem.PopupEntity(Loc.GetString("cargo-tracker-order-gone"), uid, args.User);
+    }
+
+    private void OnUseInHand(EntityUid uid, CargoTrackingComponent component, UseInHandEvent args)
+    {
+        if (component.TrackedOrderId is null)
+            return;
+
+        StopTrackingArrival(uid, args.User);
+        StopWaiting(uid, component);
+    }
+
+    private void OnComponentRemove(EntityUid uid, CargoTrackedComponent component, EntityTerminatingEvent args)
+    {
+        var station = _stationSystem.GetOwningStation(uid);
+        if (!TryComp<StationCargoOrderDatabaseComponent>(station, out var cargoOrderDatabaseComponent))
+            return;
+
+        cargoOrderDatabaseComponent.OrderLookup.Remove(component.OrderId);
+        RaiseLocalEvent(new CargoOrderDeletionEvent(component.OrderId));
+    }
+
+    private void OnCargoOrderArrival(CargoOrderArrivalEvent args)
+    {
+        // iterating over every order tracker (this is costly, but it doesn't matter because this event is raised so infrequently
+        var query = EntityQueryEnumerator<CargoTrackingComponent>();
+        while (query.MoveNext(out var uid, out var comp))
+        {
+            if (comp.TrackedOrderId == args.OrderId)
+            {
+                StartTrackingArrival(uid, args.OrderEntity, comp);
+            }
+        }
+    }
+
+    private void OnCargoOrderDeletion(CargoOrderDeletionEvent args)
+    {
+        // iterating over every order tracker (this is costly, but it doesn't matter because this event is raised so infrequently
+        var query = EntityQueryEnumerator<CargoTrackingComponent>();
+        while (query.MoveNext(out var uid, out var comp))
+        {
+            if (comp.TrackedOrderId == args.OrderId)
+            {
+                StopTrackingArrival(uid);
+            }
+        }
+    }
```

Comment:
This O(n) check is really bad.

Reply:
Have you considered keeping a mapping of the order ID to tracking entity IDs to make these infinitely faster?

Something like this? that way you can add and remove tracking items fairly quickly.
```
{ 
   "o-84723": [542352, 32462345, 23211] 
}
```

Reply:
With the power of hindsight and a good long sleep. I should have suggested you use a set instead of a list to avoid issues with double registration of entity IDs for notification